### PR TITLE
Add initial architecture document

### PR DESCRIPTION
**Pull request by cmichi**
There's things missing, of course. Like the interaction with `pallet-contracts`. I mostly want to get an initial document up and ask everyone who has something valuable to add to create a PR. This includes you too @xgreenx :blush:!

[Preview here](https://github.com/paritytech/ink/blob/cmichi-add-architecture-file/ARCHITECTURE.md).

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 These are the results when building the `examples/*` contracts from this branch with `cargo-contract 1.4.0-25bfb96` and comparing them to ink! `master`: 

|                          | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:--------------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`            |                  |            | 1.00 K               |                |
| `adder`                  |                  |            | 2.04 K               |                |
| `contract-introspection` |                  |            | 2.32 K               |                |
| `contract-terminate`     |                  |            | 0.92 K               | 275_000        |
| `contract-transfer`      |                  |            | 8.31 K               | 75_000         |
| `data-structures`        |                  |            | 1.73 K               |                |
| `delegate-calls`         |                  |            | 2.89 K               | 76_242         |
| `delegator`              |                  |            | 6.34 K               | 232_135        |
| `dns`                    |                  |            | 8.81 K               | 225_000        |
| `erc1155`                |                  |            | 16.74 K              | 450_000        |
| `erc20`                  |                  |            | 8.42 K               | 225_000        |
| `erc721`                 |                  |            | 11.62 K              | 600_000        |
| `flipper`                |                  |            | 1.24 K               | 75_000         |
| `forward-calls`          |                  |            | 2.87 K               | 151_411        |
| `incrementer`            |                  |            | 1.14 K               |                |
| `mother`                 |                  |            | 12.16 K              |                |
| `multisig`               |                  |            | 25.28 K              | 470_240        |
| `payment-channel`        |                  |            | 7.95 K               |                |
| `rand-extension`         |                  |            | 3.79 K               | 75_000         |
| `seal-code-hash`         |                  |            | 1.39 K               |                |
| `seal-ecdsa`             |                  |            | 1.71 K               |                |
| `set-code-hash`          |                  |            | 1.49 K               | 150_000        |
| `subber`                 |                  |            | 2.06 K               |                |
| `trait-erc20`            |                  |            | 8.69 K               | 225_000        |
| `trait-flipper`          |                  |            | 0.97 K               | 75_000         |
| `trait-incrementer`      |                  |            | 1.12 K               | 150_000        |
| `updated-incrementer`    |                  |            | 9.72 K               |                |
| `upgradeable-flipper`    |                  |            | 1.48 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/207350) | Last update: Wed Aug 10 18:17:18 CEST 2022

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+An important thing to note is that the crates are primarily run in
+a `no_std` environment, specifically in a Wasm sandbox execution
+environment on the blockchain itself.
+Exceptions are `metadata` and `engine`, which cover use-cases that
+are only relevant off-chain.
+
+## Overview
+
+<img src="./.images/inkternals.png" alt="ink!ternals" width="800" />
```
Comment (HCastano) [inline ARCHITECTURE.md:44]: Don't think we need to have the IR mentioned in the diagram, that's more of an implementation detail imo

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  functionality (like `vec` or `string`). Since contracts are run in a
+	`no_std` environment we provide this crate as an entrypoint for accessing
+	functionality of the standard library.
+* `primitives`: Utilities that are used internally by multiple ink! crates.
+* `storage`: The collections that are available for contract developers
+  to put in a smart contracts storage.
+
+An important thing to note is that the crates are primarily run in
+a `no_std` environment, specifically in a Wasm sandbox execution
+environment on the blockchain itself.
```
Comment (HCastano) [inline ARCHITECTURE.md:38]: This is worth expanding on. It's an important bit of context to know that the code generated will not be run on a "normal" computer like most other Rust programs

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+In the `crates/lang/` folder you'll find three separate
+crates on which `ink_lang` relies heavily:
+
+* `ink_lang_macro`: The procedural macros, they take code annotated with e.g.
+   `[ink::contract]` and forwards it to `ink_lang_ir`.
+* `ink_lang_ir`: Defines everything the procedural macro needs in order to
+   parse, analyze and generate code for ink! smart contracts.
+* `ink_lang_codegen`: Generates Rust code from the ink! IR.
+
+## Code Generation
```
Comment (HCastano) [inline ARCHITECTURE.md:61]: This is also too low level to be in an `ARCHITECTURE.md`. This should be part of developer docs within the `ink_lang_codegen` crate

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+a defined limit of space to allocate it crashes.
+This was done with the intention of reducing its complexity, which
+would have resulted in higher costs for the user (due to increased
+gas costs) and a lower transaction throughput. Freeing memory is
+irrelevant for our use-case anyway, as the entire memory instance
+is set up fresh for each individual contract call anyway.
+
+## Nightly features in ink!
+
+We would like to get away from nightly features in ink!, so that users
```
Comment (HCastano) [inline ARCHITECTURE.md:88]: I'd have a line before this giving a bit of background.

E.g: "To compile an ink! contract you need to run cargo-contract using the `nightly` compiler (`cargo +nightly ...`) ... "

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
+# ink!ternals
+
+
```
Comment (HCastano) [inline ARCHITECTURE.md:3]:
```suggestion
```

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# ink!ternals
```
Comment (HCastano) [inline ARCHITECTURE.md:1]: ink!itecture? 😂

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+You can find the crate documentation on docs.rs or for our
+`master` branch under GitHub pages. So for `ink_lang` e.g.:
+
+* [https://docs.rs/ink_lang/latest/ink_lang](https://docs.rs/ink_lang/latest/ink_lang) (latest)
+* [https://paritytech.github.io/ink/ink_lang](https://paritytech.github.io/ink/ink_lang) (`master`)
+
+ink! is composed of a number of crates that are all found in the
+`crates/` folder. On a high-level those can be grouped as:
+
+* `lang`: The ink! language itself.
```
Comment (HCastano) [inline ARCHITECTURE.md:19]: Might be worth having a diagram later showing the relationships and boundaries between the crates

File: ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+* `allocator`: The allocator for dynamic heap space allocation in a contract.
```
Comment (HCastano) [inline ARCHITECTURE.md:20]:
```suggestion
* `allocator`: The allocator used for dynamic memory allocation in a contract.
```

File: ARCHITECTURE.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+* `engine`: An off-chain testing engine, it simulates a blockchain
+   environment and allows mocking specified conditions.
+* `env`: Exposes environmental functions, like information about the caller
```
Comment (HCastano) [inline ARCHITECTURE.md:23]: Probably worth mentioning that this is the interface to the Contracts pallet as well. Could be done later in that crate diagram I proposed

File: ARCHITECTURE.md (hunk 1 above)
Comment (cmichi) [reply ARCHITECTURE.md:44]: I disagree here, I think it helps understanding what is actually happening when you build a contract.

File: ARCHITECTURE.md (hunk 2 above)
Comment (cmichi) [reply ARCHITECTURE.md:38]: Can you think of other differences besides the allocator thing?

Comment (cmichi) [reply ARCHITECTURE.md]: I added some more text on this.

File: ARCHITECTURE.md (hunk 1 above)
Comment (HCastano) [reply ARCHITECTURE.md:54]: Well it depends on what the goal here is. If we're trying to give a high level overview the fact that we have an IR is an implementation detail. You could do this without an IR and just have "raw" proc macros

File: crates/lang/codegen/src/lib.rs [hunk 10]
```diff
@@ -12,6 +12,23 @@
...
+//! was the ink! v1 way of doing code generation and it was bug prone
+//! and hard to maintain.
+//!
+//! What we do instead to generate code that is hidden from users is to
+//! pack it into anonymous `const` (so `const _: () = { … };`) and connect
+//! those to the outside world with private trait implementations.
+//!
+//! You can see how the generated code looks by installing
+//! [`cargo-expand`](https://github.com/dtolnay/cargo-expand)
+//! and executing `cd examples/flipper/ && cargo expand` in this repository.
```
Comment (HCastano) [inline crates/lang/codegen/src/lib.rs:30]: nit, but this is what I'd run instead

```suggestion
//! and executing `cargo expand --manifest-path ./examples/flipper/Cargo.toml` in this repository.
```

File: .config/cargo_spellcheck.dic [hunk 11]
```diff
@@ -62,6 +62,8 @@ stdout
 tuple
 unordered
 untyped
+v1
+v2
```
Comment (HCastano) [inline .config/cargo_spellcheck.dic:66]: I wonder if there's a way to express `v([0-9]+)` for the `.dic`

File: ARCHITECTURE.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+* [https://paritytech.github.io/ink/ink_lang](https://paritytech.github.io/ink/ink_lang) (`master`)
+
+ink! is composed of a number of crates that are all found in the
+`crates/` folder. On a high-level those can be grouped as:
+
+* `lang`: The ink! language itself.
+* `allocator`: The allocator used for dynamic memory allocation in a contract.
+* `engine`: An off-chain testing engine, it simulates a blockchain
+   environment and allows mocking specified conditions.
+* `env`: Serves two functions
```
Comment (HCastano) [inline ARCHITECTURE.md:22]:
```suggestion
* `env`: Serves two roles:
```

File: ARCHITECTURE.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+  * Exposes environmental functions, like information about the caller
+    of a contract call, getting random entropy, or e.g. self-terminating the
+    contract.
+  * Provides the connection to the [`pallet-contracts`](https://github.com/paritytech/substrate/tree/master/frame/contracts),
+    so anything that calls into the underlying execution engine of the smart contract.
+    This includes getting and setting a smart contracts storage, as well
+    as the mentioned environmental functions.
+* `metadata`: Describes the contract in a platform agnostic way, i.e.
+  its interface and the types, its storage layout, etc..
```
Comment (HCastano) [inline ARCHITECTURE.md:31]:
```suggestion
  its interface and the types, its storage layout, etc.
```

**Review by HCastano — APPROVED** [review]
I still disagree about the IR thing, but this is a good starting doc. Thanks!

**Pull request merged**